build: add size-tracking bazel test

devversion opened the pull request:
Introduces a new Bazel test that allows us to inspect
what source-files contribute to a given bundled file and
how much each file contributes to the bundle size in bytes.

Additionally the size-tracking rule groups the size
data by directories. This allows us to compare size
changes in the scope of directories. e.g. a lot of
files in a directory could increase slightly in size, but
in the directory scope the size change could be significant
and needs to be reported by the test target.

Resolves FW-1278

devversion committed:
build: add size-tracking bazel test

Introduces a new Bazel test that allows us to inspect
what source-files contribute to a given bundled file
and how much bytes they contribute to the bundle size.

Additionally the size-tracking rule groups the size
data by directories. This allows us to compare size
changes in the scope of directories. e.g. a lot of
files in a directory could increase slightly in size, but
in the directory scope the size change could be significant
and needs to be reported by the test target.

Resolves FW-1278

mhevery commented on packages/core/test/bundling/core_all/bundle.golden_size_map.json:
```diff
@@ -0,0 +1,267 @@
+{
+  "unmapped": 25,
+  "files": {
```
1.) Could we merge the files and directories like this
```
  "src/render3/i18n.ts": 1234,
  "src/render3/": 1234,  // See trailing slash
```

2.) Could we get the JSON keys to be alphabetical?

mhevery replied:
Actually I thinks this format would be more useful to the developer:

```
{
 " ": 1234,
  "src": {
    " ": 1234
    "render3": {
      " ": 1234,
      "someFile.ts": 1234
    },
    "di": {
      " ": 1234
      "someFile.ts": 1234
     }
  }
}
```

I think this JSON format will give instant tree like insight for the developer. What do you think?

devversion replied:
Re (1): Good idea with merging them into one object. Also the trailing slash is good for implying that this is a directory :+1:

Re (2): I initially had that but thought it would be more helpful to have it sorted by size. That way it's easy to see what contributes the most to the bundle. I'm fine changing to what you prefer.

Regarding the hierarchy: I also did played with that initially but it resulted in a more confusing golden file due to the overhead and long external paths with a lot of nesting. Up to you. I totally see the point of making it a tree that reflects the actual file system.

mhevery replied:
I think we should reflect the file system. Not suer that sorting by size is of much value.

devversion replied:
Done as discussed.

devversion committed:
fixup! build: add size-tracking bazel test

Update golden to non-Windows bundle. Windows is currently inconsistent due to a bug that will be fixed with: https://github.com/bazelbuild/rules_nodejs/pull/720

mhevery commented on tools/size-tracking/BUILD.bazel:
```diff
@@ -0,0 +1,13 @@
+        "@npm//@types/node",
+        "@npm//@types/source-map",
+    ],
+)
```
Where are the specs for the `size-tracking` build rule?

devversion replied:
Pushed them now.

mhevery commented on tools/size-tracking/file_size_data.ts:
```diff
@@ -0,0 +1,33 @@
+
+function _sortEntrySizeObject(oldObject: SizeMap, newObject: SizeMap) {
+  Object.keys(oldObject)
+      .sort((a, b) => oldObject[b] - oldObject[a])
```
Hmm, I see that you are sorting but the gold file does not seem to be sorted.

devversion replied:
The golden file is compared by value and the order does not matter.

mhevery commented on tools/size-tracking/index.ts:
```diff
@@ -0,0 +1,56 @@
+if (require.main === module) {
+  const [filePath, sourceMapPath, goldenPath, thresholdArg, writeGoldenArg] = process.argv.slice(2);
+  const status = main(
+      require.resolve(filePath), require.resolve(sourceMapPath), require.resolve(goldenPath),
```
WOW, that is a good trick which I did not think of. 👍

devversion commented on packages/core/test/bundling/core_all/BUILD.bazel:
```diff
@@ -0,0 +1,41 @@
+        ":bundle",
+        ":bundle.golden_size_map.json",
+    ],
+    diffThreshold = 2,
```
@mhevery We need to think about the threshold we want to use. Also whether the difference percentage is the best way of keeping track of size regressions. I could also imagine just checking against a fixed byte-threshold.

mhevery replied:
Open to ideas. I don't really have a feel for which strategy would be best here.

devversion replied:
I'm not entirely sure either. It depends on how the tool is used. if we want to make sure that size changes which impact the bundle are "approved" per file then the percentage strategy is probably better as it is quite strict.

Though I'm a bit worried that we report failures too often with the percentage difference strategy as _super small_ changes to small files cause the difference percentage to exceed the threshold quickly. e.g. `100byte <> 110byte > 9.5% (10 byte) change`. And for large files, a lot of changes are required. e.g. `1000 > 1100 > 9.5% (100 byte)`.

If we have a fixed byte threshold, then it would work better for small changes, but it would report also report a size-change if a given file is already quite large but a relatively small change has been made.

mhevery replied:
Could we have both. Something like max size and min percentage?

devversion replied:
@mhevery I like the idea of having both. I'll do that in a follow-up since the test is disabled anyway for now.

mhevery commented on packages/core/test/bundling/core_all/bundle.golden_size_map.json:
```diff
@@ -0,0 +1,816 @@
+        "size": 247422,
+        "src/": {
+          "size": 247341,
+          "application_init.ts": {
```
if it is a file can we just have value rather than object literal (It is just noise).

```suggestion
          "application_init.ts": 626
```

mhevery commented:
LGTM, but can you change the output per comment, otherwise this is great.

devversion committed:
fixup! build: add size-tracking bazel test

Add size-tracking tests and store golden in hierarchical tree to reflect file system

kara commented:
merge-assistance: global approval

IgorMinar reviewed (commented):
isn't this going to make the CI super brittle because the numbers will fluctuate all the time?

isn't what we want something that will monitor the numbers but not break the CI? all we need is ability to understand what the breakdown is and how it has changed over time (and what PRs/commits) to attribute the changes to.

IgorMinar commented:
let's discuss this more before merging

mhevery commented:
Discussed this with @IgorMinar and we agreed to:

- Merge this in as is, to see how much paint this generates for people.
- If the pain is too high, then we disable the test on the CI so that we run it manually only.

mhevery committed:
fixup! build: add size-tracking bazel test

devversion committed:
fixup! build: add size-tracking bazel test

Update golden to be up-to-date with master

devversion commented:
@AndrewKushnir Rebased. For later reference. The test will not run until we remove the `manual` tag. This is by intention because we still want to decide how we handle the failures/size assertions after NgConf.
